pkg/controllers/interruption/controller.go

```diff
@@ -0,0 +1,275 @@
+// assigned action on the instances based on the message event
+func (c *Controller) handleMessage(ctx context.Context, instanceIDMap map[string]*v1.Node, raw *sqsapi.Message) error {
+	// No message to parse in this case
+	if raw == nil || raw.Body == nil {
+		return nil
+	}
+	msg, err := c.parser.Parse(*raw.Body)
+	if err != nil {
+		// In the scenario where we can't parse the message, we log that we have an error and then are
+		// forced to just delete the message from the queue
+		logging.FromContext(ctx).Errorf("parsing sqs message, %v", err)
+		err = c.provider.DeleteSQSMessage(ctx, raw)
+		if err != nil {
+			return fmt.Errorf("failed to delete message from queue, %w", err)
+		}
+		deletedMessages.Inc()
+		return nil
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("event", msg.Kind()))
+
+	nodes := getInvolvedNodes(msg.EC2InstanceIDs(), instanceIDMap)
+	// There's no action to take here since the event doesn't pertain to any of our instances
+	if len(nodes) == 0 {
+		receivedMessages.WithLabelValues(msg.Kind().String(), "false").Inc()
+
+		// Since there's no action, just delete the message
+		err = c.provider.DeleteSQSMessage(ctx, raw)
+		if err != nil {
+			return fmt.Errorf("failed to delete message from queue, %w", err)
+		}
+		deletedMessages.Inc()
+		return nil
+	}
+	receivedMessages.WithLabelValues(msg.Kind().String(), "true").Inc()
+
+	nodeNames := lo.Map(nodes, func(n *v1.Node, _ int) string { return n.Name })
+	logging.FromContext(ctx).Infof("Received actionable event from SQS queue for node(s) [%s%s]",
+		strings.Join(lo.Slice(nodeNames, 0, 3), ","),
+		lo.Ternary(len(nodeNames) > 3, "...", ""))
+
+	for i := range nodes {
+		node := nodes[i]
+		err = multierr.Append(err, c.handleNode(ctx, msg, node))
+	}
+	if err != nil {
+		return fmt.Errorf("failed to act on nodes [%s%s], %w",
+			strings.Join(lo.Slice(nodeNames, 0, 3), ","),
+			lo.Ternary(len(nodeNames) > 3, "...", ""), err)
+	}
+	err = c.provider.DeleteSQSMessage(ctx, raw)
+	if err != nil {
+		return fmt.Errorf("failed to delete message from queue, %w", err)
+	}
+	deletedMessages.Inc()
+	return nil
+}
+
+func (c *Controller) handleNode(ctx context.Context, msg messages.Message, node *v1.Node) error {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("node", node.Name))
+	action := actionForMessage(msg)
+
+	// Record metric and event for this action
+	c.notifyForMessage(msg, node)
+	actionsPerformed.WithLabelValues(action.String()).Inc()
+
+	// Mark the offering as unavailable in the ICE cache since we got a spot interruption warning
+	if msg.Kind() == messages.SpotInterruptionKind {
+		zone := node.Labels[v1.LabelTopologyZone]
+		instanceType := node.Labels[v1.LabelInstanceTypeStable]
+		if zone != "" && instanceType != "" {
+			c.unavailableOfferingsCache.MarkUnavailable(ctx, msg.Kind().String(), instanceType, zone, v1alpha1.CapacityTypeSpot)
+		}
+	}
+	if action != NoAction {
```

Comment:
Thoughts on simply not having a "NoAction" message? Seems like we could model it more implicitly, such that there simply aren't any actions to take from the message.

Reply:
I think I'm ok with it since the action abstraction is encapsulating the behavior of what to do based on the message type. I don't want to leak those details into the reconcile loop